src/app/courses/course-editor/course-editor.component.ts

```diff
@@ -0,0 +1,97 @@
+import { Component, OnInit } from '@angular/core';
+import { switchMap } from 'rxjs/operators';
+import { of, Observable } from 'rxjs';
+
+import { Course } from '../course.model';
+import { ActivatedRoute, ParamMap, Router } from '@angular/router';
+import { CourseService } from '../services/course.service';
+import { CanComponentDeactivate } from 'src/app/core/guards/can-deactivate.guard';
+
+@Component({
+  selector: 'app-course-editor',
+  templateUrl: './course-editor.component.html',
+  styleUrls: ['./course-editor.component.scss']
+})
+export class CourseEditorComponent implements OnInit, CanComponentDeactivate {
+  public newCourse: Course;
+  public editedCourse: Course;
+
+  constructor(
+    private route: ActivatedRoute,
+    private router: Router,
+    private courseService: CourseService
+  ) { }
+
+  ngOnInit() {
+    this.editedCourse = null;
+    this.newCourse = this.createNewCourse();
+    this.route.paramMap.pipe(
+      switchMap((params: ParamMap) =>
+        of(this.courseService.getCourseById(+params.get('id')))))
+        .subscribe((course: Course) => {
+          if (course) {
+            this.editedCourse = course;
+            this.newCourse = { ...course };
+          }
+        });
+  }
+
+  public get authors(): string {
+    return this.newCourse.authors.join(', ');
+  }
+
+  public set authors(value: string) {
+    this.newCourse.authors = value.split(',').map((item: string) => item.trim());
+  }
+
+  public saveCourse(): void {
+    console.log('Changes saved successfully.');
+    console.log(this.newCourse);
+    this.editedCourse ?
+      this.courseService.updateCourse(this.editedCourse, this.newCourse) :
+      this.courseService.createCourse(this.newCourse);
+
+     this.router.navigate(['/courses']);
+  }
+
+  public returnToCourses(): void {
+    this.router.navigate(['/courses']);
+  }
+
+  public setNewDate(newDate: string): void {
+    console.log(newDate);
+    this.newCourse.creationDate = newDate;
+  }
+
+  public setDuration(newDuration: string): void {
+    console.log(newDuration);
+    this.newCourse.duration = newDuration;
+  }
+
+  public createNewCourse(): Course {
+    return {
+      id: this.generateCourseId(),
```

Comment:
You could use Course class with default parameters. In course.model.ts
export class Course implements CoursesListItem {
  constructor(
    public id = null,
    public title = '',
    public creationDate = '',
    public duration = '',
    public description = '',
    public authors = [],
    public topRated = false
  ) {}
}